tools/tpm2_hierarchycontrol.c

```diff
@@ -0,0 +1,174 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+#include <string.h>
+
+#include <log.h>
+#include <tpm2.h>
+#include <tpm2_capability.h>
+#include "tpm2_tool.h"
+
+typedef struct hierarchycontrol_ctx hierarchycontrol_ctx;
+struct hierarchycontrol_ctx {
+    struct {
+        const char *ctx_path;
+        const char *auth_str;
+        tpm2_loaded_object object;
+    } auth_hierarchy;
+    TPMI_RH_ENABLES enable;
+    TPMI_YES_NO state;
+};
+
+static hierarchycontrol_ctx ctx = {
+    .auth_hierarchy.ctx_path="p",
+};
+
+static tool_rc hierarchycontrol(ESYS_CONTEXT *ectx) {
+
+    LOG_INFO ("Sending TPM2_HierarchyControl bit(%s) \'%s\' command with auth handle %s",
+        ctx.enable == TPM2_RH_PLATFORM ? "phEnable" :
+        ctx.enable == TPM2_RH_OWNER ? "shEnable" :
+        ctx.enable == TPM2_RH_ENDORSEMENT ? "ehEnable" : "phEnableNV",
+        ctx.state ? "SET" : "CLEAR",
+        ctx.auth_hierarchy.object.tr_handle == ESYS_TR_RH_OWNER ? "TPM2_RH_OWNER" :
+        ctx.auth_hierarchy.object.tr_handle == ESYS_TR_RH_ENDORSEMENT ? "TPM2_RH_ENDORSEMENT" :
+        ctx.auth_hierarchy.object.tr_handle == ESYS_TR_RH_PLATFORM ? "TPM2_RH_PLATFORM" : "TPM2_RH_PLATFORM_NV");
+
+    return tpm2_hierarchycontrol(ectx, &ctx.auth_hierarchy.object, ctx.enable, ctx.state);
+}
+
+bool on_arg (int argc, char **argv) {
+
+    if (!argc)
+        goto error;
+
+    if (argc > 1)
+        goto error;
+
+    if (!strcmp(argv[0], "set")) {
+        ctx.state = TPM2_YES;
+	return true;
+    }
+
+    if (!strcmp(argv[0], "clear")) {
+        ctx.state = TPM2_NO;
+	return true;
+    }
+
+error:
+    LOG_ERR("Incorrect operation value, got: \"%s\", expected [set|clear]", argv[0]);
+    return false;
+}
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'c':
+        ctx.auth_hierarchy.ctx_path = value;
+        break;
+    case 'e':
+        if (!strcmp(value, "phEnable")) {
+            ctx.enable = TPM2_RH_PLATFORM;
+        } else if (!strcmp(value, "shEnable")) {
+            ctx.enable = TPM2_RH_OWNER;
+        } else if (!strcmp(value, "ehEnable")) {
+            ctx.enable = TPM2_RH_ENDORSEMENT;
+        } else if (!strcmp(value, "phEnableNV")) {
+            ctx.enable = TPM2_RH_PLATFORM_NV;
+        } else {
+            LOG_ERR("Incorrect property, got: \"%s\", expected [phEnable|shEnable|ehEnable|phEnableNV]", value);
+            return false;
+        }
+        break;
+    case 'p':
+        ctx.auth_hierarchy.auth_str = value;
+        break;
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    const struct option topts[] = {
+        { "context-object",	required_argument, NULL, 'c' },
+        { "auth-key",		required_argument, NULL, 'p' },
+        { "enable",		required_argument, NULL, 'e' },
```

Comment:
delete

Reply:
What did mean "delete"?
Delete these 4 lines or modify them to correct option?

Reply:
I think @williamcroberts is referring to his comment above

> kill this option and make it argument.
> So
> ```sh
> tpm2_hierarchycontrol -c e -e phEnable set
> ```
> becomes:
> ```sh
> tpm2_hierarchycontrol -c e phEnable set
> ```

As a consequence there would not be an option `-e`/`--enable` any more.

Reply:
Yeah just delete the -e line and move it into the tools arguments. Sorry for the thrash on the review, we are scrutinizing our options as we move into the 4.0 release with vigor, as we are stuck with those choices. As we move forward, we will stabilize the CLI.

Reply:
No problem, please review patches again, Thanks.